processor/metricstransformprocessor/metrics_transform_processor.go

```diff
@@ -0,0 +1,195 @@
+	"go.opentelemetry.io/collector/component"
+	"go.opentelemetry.io/collector/consumer"
+	"go.opentelemetry.io/collector/consumer/pdata"
+	"go.opentelemetry.io/collector/consumer/pdatautil"
+)
+
+const (
+	validNewNameFuncName  = "validNewName()"
+	validNewLabelFuncName = "validNewLabel()"
+)
+
+type metricsTransformProcessor struct {
+	cfg        *Config
+	next       consumer.MetricsConsumer
+	transforms []Transform
+}
+
+var _ component.MetricsProcessor = (*metricsTransformProcessor)(nil)
+
+func newMetricsTransformProcessor(next consumer.MetricsConsumer, cfg *Config) (*metricsTransformProcessor, error) {
+	return &metricsTransformProcessor{
+		cfg:        cfg,
+		next:       next,
+		transforms: cfg.Transforms,
+	}, nil
+}
+
+// GetCapabilities returns the Capabilities associated with the metrics transform processor.
+func (mtp *metricsTransformProcessor) GetCapabilities() component.ProcessorCapabilities {
+	return component.ProcessorCapabilities{MutatesConsumedData: true}
+}
+
+// Start is invoked during service startup.
+func (*metricsTransformProcessor) Start(ctx context.Context, host component.Host) error {
+	return nil
+}
+
+// Shutdown is invoked during service shutdown.
+func (*metricsTransformProcessor) Shutdown(ctx context.Context) error {
+	return nil
+}
+
+// ConsumeMetrics implements the MetricsProcessor interface.
+func (mtp *metricsTransformProcessor) ConsumeMetrics(ctx context.Context, md pdata.Metrics) error {
+	return mtp.next.ConsumeMetrics(ctx, mtp.transform(md))
+}
+
+// transform transforms the metrics based on the information specified in the config.
+func (mtp *metricsTransformProcessor) transform(md pdata.Metrics) pdata.Metrics {
+	mds := pdatautil.MetricsToMetricsData(md)
+
+	for i, data := range mds {
+		nameToMetricMapping := make(map[string]*metricspb.Metric)
+		// O(len(data.Metrics))
+		for _, metric := range data.Metrics {
+			nameToMetricMapping[metric.MetricDescriptor.Name] = metric
+		}
+
+		for _, transform := range mtp.transforms {
+			if !mtp.validNewName(transform, nameToMetricMapping) {
+				log.Printf("error running %q processor due to collision %q: %v with existing metric names detected by the function %q", typeStr, NewNameFieldName, transform.NewName, validNewNameFuncName)
+				continue
+			}
+
+			metric, ok := nameToMetricMapping[transform.MetricName]
+			if !ok {
+				continue
+			}
+
+			// mtp.action is already validated to only contain either update or insert
+			if transform.Action == Update {
+				mtp.update(metric, transform)
+				if transform.NewName == "" {
+					continue
+				}
+				// if name is updated, the map has to be updated
+				nameToMetricMapping[transform.NewName] = nameToMetricMapping[transform.MetricName]
+				delete(nameToMetricMapping, transform.MetricName)
+			} else if transform.Action == Insert {
+				var newMetric *metricspb.Metric
+				mds[i].Metrics, newMetric = mtp.insert(metric, mds[i].Metrics, transform)
```

Comment:
Nit: you can put `mds[i]` into a variable so you don't need to dereference it on each iteration

Reply:
I actually tried to use `data` directly, which is `mds[i]` from the for loop, but this way, the original copy of the metric is not updated correctly since mds is a list of non-pointers.

Reply:
Oh interesting. I guess you could set `data := &mds[i]` instead